### DISCUSSION TITLE
Should entries representing constructors in the component browser be capitalized?

### DISCUSSION THREAD

**Discussion by radeusgd** [enso-org/enso#6568] [Work In Progress - GUI] [answered]
I will just repost [my response from today's bookclub](https://discord.com/channels/401396655599124480/1103622151753441340/1103734795013472387), because it is important to me and I feel like it may get lost on Discord. Also discussion will be cleaner in a separate place.

--------------------------

It is also very confusing to me that the entries corresponding to Enso types and constructors which are written as **`First_Letter_Capitalized`** show up **lowercase** in the CB.

Not only this is inconsistent - searcher shows me lowercase but once I insert it it becomes (correctly!) uppercase - so the code inserted is different from what searcher showed.

But also it makes searching harder for me. I'm already used to seeing these capitalized names, having worked a lot with them in Text mode (and I imagine any GUI user will be too, as in the final code / on nodes they **will** show capitalized). But the searcher shows them lowercase - which is a sight that I'm **not** used to. So it makes the experience a bit uncomfortable because I need to translate in my head between lower and upper case.

Not only that, but it makes it harder to distinguish methods (lowercase) from constructors (capitalied) in the CB. Yes, I see some rhombus vs circle icons - but these tell me nothing -  I need to learn and remember which icon means what and I'm not at that level of GUI expertise yet. Whereas the lowercase/capitalized distinction is very well known in the Enso world, so I don't understand why not just rely on this distinction here as well.

----

Picture showing **capitalized** `Sum` _constructor_ displayed as lowercase `sum` (looking just like a regular _method_) in the CB:
![image](https://user-images.githubusercontent.com/1436948/236282857-706c9a63-b0b0-42e8-b14b-f668e9477111.png)

**Comment by wdanilo** [comment]
I didn't know we are showing lowercase letters in the searcher always. I agree this can be confusing. On the other hand though .... if you think about the whole new node design, like this:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/1623053/236291230-8d72b899-7bf4-4336-9c91-c25623eb1ef0.png">

Then the expression on this node is basically `... as=(Std.Something.Delimited Std.Something.Dot) ...`. Using upper-case letters there would make the node WAY less user friendly and look strange. I understand that this is not a code-view and it diverges from the code view more and more, but this is the whole purpose of widgets to be very natural to users who are non-programmers. We could make it upper-case though, and maybe we should. I will create a few tests in Figma and I'll report back, but I'd also love to hear your thoughts on that ^^^

**Comment by farmaazon** [comment]
The Component Browser was designed to be less a "completion" but more "node browser" tool, so it shows the demangled name of the node. It's quite expected that code will differ from the displayed in CB. Also, I wonder if user will need to pick constructors from the Component Browser when widgets will be available? Most constructors will be inserted by drop-down widget.

**Reply by wdanilo** [reply]
It should be possible for the user to use constructors to construct data types and pass them down the lines in the graph. E.g. Creating a `Point_2d` and then doing something with it. In such cases, distinction between upper and Lower-case letters is important. I need to think about it more – this is tricky. Let's catch up on Monday, please (in case this thread dies, let's ping ourselves, please).

**Reply by radeusgd** [reply]
@wdanilo Pinging as the discussion seems to have died without resolution.

**Reply by wdanilo** [reply]
@radeusgd thanks a lot. I'm creating a call about it for Monday. Is it OK or do you need to chat earlier about it?

**Comment by farmaazon** [comment]
During the call with @wdanilo and @radeusgd we decided to keep all things lowercase in the widgets; when a widget is edited as "Enso code", the user will see constructors as in code, i.e. uppercase. When we implement it, we'll see if it's confusing and decide about next steps (which may include changing Enso language to use lowercase only).

Regarding CB, here we should make a distinction between "code editing mode" and "adding node mode":
1. The first mode is used when the user edits a single widget or an entire node as a code, or when they add a new node as a code (a different shortcut than adding a node normally). In that case, the CB should display all entries without demangling (with `_` and preserving case). The CB is context-aware, so when user write "`read `" (mark the space), the CB will show suggestions for first argument of `read` method.
2. The second mode is used when the user adds a node normally. The suggestions are demangled - `_` are replaced with ` ` and lower-cased. The entire input is treated as a filtering pattern, without any contexts, so when user write "`read `", the CB will show suggestions filtered by "`read `", for example "read text".

Currently, we have a mix of both modes, where suggestions on list are demangled, but input is context-aware. I will create an issue for new modes soon (as part of #6622 probably).

**Comment by farmaazon** [comment, marked as answer]
After all, another answer reached [the new design](https://github.com/enso-org/design/blob/main/epics/component-browser/design2.md) - the widgets and CB should display Enso code without changing case.
